File: src/app/observability.ts
/**
 * Vendor-neutral diagnostics for the web client.
 *
 * This module is deliberately a local boundary. It does not send anything,
 * read workspace data, or decide consent/retention. A future approved sink can
 * consume the already-redacted event shape without receiving raw exceptions,
 * request bodies, or account identifiers.
 */

export const ROOTINE_OBSERVABILITY_SCHEMA_VERSION = 1 as const;
export const ROOTINE_DIAGNOSTIC_EVENT_MAX_BYTES = 4_096;
export const ROOTINE_DIAGNOSTIC_EXPORT_MAX_BYTES = 64 * 1_024;
export const ROOTINE_DIAGNOSTIC_EVENT_LIMIT = 64;

export const ROOTINE_TELEMETRY_EVENT_NAMES = [
  "auth_outcome",
  "sync_operation",
  "realtime_health",
  "qr_scan",
  "notification_delivery",
  "materializer_quarantine",
  "device_health",
  "crash",
  "support_export",
] as const;
export type RootineTelemetryEventName = (typeof ROOTINE_TELEMETRY_EVENT_NAMES)[number];

export const ROOTINE_HEALTH_COUNTER_NAMES = [
  "auth_success",
  "auth_failure",
  "sync_pull_success",
  "sync_pull_failure",
  "sync_push_success",
  "sync_push_failure",
  "sync_retry",
  "sync_conflict",
  "sync_cursor_expired",
  "sync_unauthorized",
  "realtime_connected",
  "realtime_reconnect",
  "realtime_failure",
  "qr_detected",
  "qr_success",
  "qr_failure",
  "apns_delivered",
  "apns_failed",
  "apns_unregistered",
  "apns_retry",
  "materializer_quarantine",
  "device_registered",
  "device_registration_failure",
  "crash_captured",
] as const;
export type RootineHealthCounterName = (typeof ROOTINE_HEALTH_COUNTER_NAMES)[number];

export type RootineTelemetryOutcome = "success" | "failure" | "degraded" | "cancelled" | "unknown";

export type RootineDiagnosticEvent = {
  schemaVersion: typeof ROOTINE_OBSERVABILITY_SCHEMA_VERSION;
  name: RootineTelemetryEventName;
  outcome: RootineTelemetryOutcome;
  at: string;
  durationMs?: number;
  correlationId?: string;
  operationId?: string;
  attributes: Record<string, string | number | boolean>;
};

export type RootineDiagnosticSnapshot = {
  schemaVersion: typeof ROOTINE_OBSERVABILITY_SCHEMA_VERSION;
  generatedAt: string;
  supportId: string;
  counters: Partial<Record<RootineHealthCounterName, number>>;
  events: RootineDiagnosticEvent[];
};

export type RootineTelemetrySink = {
  record: (event: RootineDiagnosticEvent) => void;
};

const PRIVATE_KEY = /(?:access.?token|authorization|cookie|password|secret|push.?token|token|payload|record|notes?|health|finance|financial|nutrition|content|body|text|title|description|location|latitude|longitude|email)/i;
const SAFE_ATTRIBUTE_KEYS = new Set([
  "action",
  "endpoint",
  "environment",
  "error",
  "entity",
  "entityId",
  "format",
  "httpStatus",
  "permission",
  "provider",
  "reason",
  "source",
  "status",
  "trigger",
  "attempt",
  "batchSize",
  "changeCount",
  "cursor",
  "revision",
  "retryAfterSeconds",
  "queueDepth",
  "sampleRate",
]);

function boundedString(value: string, max = 180) {
  const normalized = Array.from(value, (character) => {
    const code = character.charCodeAt(0);
    return code < 32 || code === 127 ? " " : character;
  }).join("").trim();
  return normalized.length > max ? `${normalized.slice(0, max - 3)}...` : normalized;
}

function errorCode(value: unknown) {
  const normalized = String(value ?? "").toLocaleLowerCase("en-US");
  if (normalized.includes("unauthor") || normalized.includes("401")) return "unauthorized";
  if (normalized.includes("cursor") && normalized.includes("expir")) return "cursor_expired";
  if (normalized.includes("conflict") || normalized.includes("revision")) return "conflict";
  if (normalized.includes("rate") || normalized.includes("429")) return "rate_limited";
  if (normalized.includes("timeout") || normalized.includes("timed out") || normalized.includes("abort")) return "timeout";
  if (normalized.includes("network") || normalized.includes("fetch")) return "network";
  if (normalized.includes("invalid") || normalized.includes("schema")) return "invalid";
  if (normalized.includes("server") || normalized.includes("5")) return "server";
  return "unknown";
}

function safeIdentifier(value: unknown, max = 180) {
  return typeof value === "string" && value.trim() ? boundedString(value, max) : undefined;
}

/** Redacts arbitrary diagnostic values before they can reach a future sink. */
export function redactRootineDiagnosticValue(value: unknown, key?: string): unknown {
  if (key && PRIVATE_KEY.test(key)) return "[REDACTED]";
  if (typeof value === "string") return boundedString(value);
  if (typeof value === "number" || typeof value === "boolean" || value === null) return value;
  if (Array.isArray(value)) return value.slice(0, 16).map((item) => redactRootineDiagnosticValue(item));
  if (!value || typeof value !== "object") return undefined;
  return Object.fromEntries(
    Object.entries(value as Record<string, unknown>)
      .filter(([entryKey]) => !PRIVATE_KEY.test(entryKey))
      .slice(0, 32)
      .map(([entryKey, entryValue]) => [entryKey, redactRootineDiagnosticValue(entryValue, entryKey)]),
  );
}

function safeAttributes(attributes: Record<string, unknown> | undefined) {
  if (!attributes) return {};
  return Object.fromEntries(
    Object.entries(attributes)
      .filter(([key]) => SAFE_ATTRIBUTE_KEYS.has(key))
      .slice(0, 24)
      .map(([key, value]) => {
        if (key === "error" || key === "reason") return [key, errorCode(value)];
        if (typeof value === "number" && Number.isFinite(value)) return [key, Math.round(value)];
        if (typeof value === "boolean") return [key, value];
        if (typeof value === "string") return [key, boundedString(value)];
        return [key, String(redactRootineDiagnosticValue(value, key) ?? "unknown")];
      }),
  ) as Record<string, string | number | boolean>;
}

function serializedBytes(value: unknown) {
  return new TextEncoder().encode(JSON.stringify(value)).byteLength;
}

function fallbackSupportId() {
  return `web-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
}

function newSupportId() {
  try {
    if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
      return `web-${crypto.randomUUID()}`;
    }
  } catch {
    // A diagnostic identifier must never make an app flow fail.
  }
  return fallbackSupportId();
}

function boundedEvent(input: {
  name: RootineTelemetryEventName;
  outcome?: RootineTelemetryOutcome;
  at?: string;
  durationMs?: number;
  correlationId?: string;
  operationId?: string;
  attributes?: Record<string, unknown>;
}): RootineDiagnosticEvent {
  const event: RootineDiagnosticEvent = {
    schemaVersion: ROOTINE_OBSERVABILITY_SCHEMA_VERSION,
    name: input.name,
    outcome: input.outcome ?? "unknown",
    at: typeof input.at === "string" ? boundedString(input.at, 48) : new Date().toISOString(),
    ...(Number.isFinite(input.durationMs) ? { durationMs: Math.max(0, Math.round(input.durationMs!)) } : {}),
    ...(safeIdentifier(input.correlationId) ? { correlationId: safeIdentifier(input.correlationId) } : {}),
    ...(safeIdentifier(input.operationId) ? { operationId: safeIdentifier(input.operationId) } : {}),
    attributes: safeAttributes(input.attributes),
  };
  if (serializedBytes(event) <= ROOTINE_DIAGNOSTIC_EVENT_MAX_BYTES) return event;
  const trimmed = { ...event, attributes: {} };
  for (const [key, value] of Object.entries(event.attributes)) {
    const candidate = { ...trimmed, attributes: { ...trimmed.attributes, [key]: value } };
    if (serializedBytes(candidate) > ROOTINE_DIAGNOSTIC_EVENT_MAX_BYTES) break;
    trimmed.attributes = candidate.attributes;
  }
  return trimmed;
}

export class RootineObservability {
  static readonly shared = new RootineObservability();

  private readonly sink?: RootineTelemetrySink;
  private readonly now: () => Date;
  private readonly maxEvents: number;
  private readonly supportId: string;
  private readonly counters = new Map<RootineHealthCounterName, number>();
  private events: RootineDiagnosticEvent[] = [];

  constructor(options: {
    sink?: RootineTelemetrySink;
    now?: () => Date;
    supportId?: string;
    maxEvents?: number;
  } = {}) {
    this.sink = options.sink;
    this.now = options.now ?? (() => new Date());
    const requestedEventLimit = Number.isFinite(options.maxEvents) ? Math.floor(options.maxEvents!) : ROOTINE_DIAGNOSTIC_EVENT_LIMIT;
    this.maxEvents = Math.max(1, Math.min(ROOTINE_DIAGNOSTIC_EVENT_LIMIT, requestedEventLimit));
    this.supportId = safeIdentifier(options.supportId, 96) ?? newSupportId();
  }

  get identifier() {
    return this.supportId;
  }

  record(input: Omit<Parameters<typeof boundedEvent>[0], "at"> & { at?: string }) {
    const event = boundedEvent({ ...input, at: input.at ?? this.now().toISOString() });
    this.events = [...this.events, event].slice(-this.maxEvents);
    try {
      this.sink?.record(event);
    } catch {
      // A future sink is optional and must never break a product flow.
    }
    return event;
  }

  increment(name: RootineHealthCounterName, amount = 1) {
    if (!Number.isFinite(amount) || amount <= 0) return;
    const next = (this.counters.get(name) ?? 0) + Math.min(1_000, Math.round(amount));
    this.counters.set(name, Math.min(Number.MAX_SAFE_INTEGER, next));
  }

  recordAuthOutcome(input: { action: string; outcome: RootineTelemetryOutcome; provider?: string; error?: unknown }) {
    this.increment(input.outcome === "success" ? "auth_success" : "auth_failure");
    const attributes = {
      action: input.action,
      provider: input.provider ?? "password",
      ...(input.error === undefined ? {} : { error: input.error }),
    };
    return this.record({
      name: "auth_outcome",
      outcome: input.outcome,
      attributes,
    });
  }

  recordSyncOperation(input: {
    endpoint: "bootstrap" | "pull" | "push" | "register_device" | string;
    outcome: RootineTelemetryOutcome;
    status?: string;
    durationMs?: number;
    correlationId?: string;
    operationId?: string;
    error?: unknown;
    attributes?: Record<string, unknown>;
  }) {
    const prefix = input.endpoint === "pull" ? "sync_pull" : input.endpoint === "push" ? "sync_push" : undefined;
    if (prefix) this.increment(`${prefix}_${input.outcome === "success" ? "success" : "failure"}` as RootineHealthCounterName);
    const status = String(input.status ?? "").toLocaleLowerCase("en-US");
    if (status === "conflict") this.increment("sync_conflict");
    if (status === "cursor_expired") this.increment("sync_cursor_expired");
    if (status === "unauthorized") this.increment("sync_unauthorized");
    if (status === "retry" || status === "rate_limited") this.increment("sync_retry");
    return this.record({
      name: "sync_operation",
      outcome: input.outcome,
      durationMs: input.durationMs,
      correlationId: input.correlationId,
      operationId: input.operationId,
      attributes: { endpoint: input.endpoint, status: input.status, error: input.error, ...input.attributes },
    });
  }

  recordQrScan(input: { outcome: "detected" | "success" | "failure"; format?: "qr" | "barcode" | string; error?: unknown }) {
    this.increment(input.outcome === "detected" ? "qr_detected" : input.outcome === "success" ? "qr_success" : "qr_failure");
    const attributes = {
      format: input.format === "qr" || input.format === "barcode" ? input.format : "unknown",
      ...(input.error === undefined ? {} : { error: input.error }),
    };
    return this.record({
      name: "qr_scan",
      outcome: input.outcome === "failure" ? "failure" : "success",
      attributes,
    });
  }

  recordNotificationDelivery(input: { status: "delivered" | "failed" | "unregistered" | "retry"; environment?: string; error?: unknown }) {
    const counter = input.status === "delivered" ? "apns_delivered" : input.status === "unregistered" ? "apns_unregistered" : input.status === "retry" ? "apns_retry" : "apns_failed";
    this.increment(counter);
    return this.record({ name: "notification_delivery", outcome: input.status === "delivered" ? "success" : input.status === "retry" ? "degraded" : "failure", attributes: {
      status: input.status,
      ...(input.environment === undefined ? {} : { environment: input.environment }),
      ...(input.error === undefined ? {} : { error: input.error }),
    } });
  }

  recordMaterializerQuarantine(reason?: string) {
    this.increment("materializer_quarantine");
    return this.record({ name: "materializer_quarantine", outcome: "degraded", attributes: reason === undefined ? {} : { reason } });
  }

  recordCrash(error?: unknown) {
    this.increment("crash_captured");
    return this.record({ name: "crash", outcome: "failure", attributes: error === undefined ? {} : { error } });
  }

  snapshot(): RootineDiagnosticSnapshot {
    return {
      schemaVersion: ROOTINE_OBSERVABILITY_SCHEMA_VERSION,
      generatedAt: this.now().toISOString(),
      supportId: this.supportId,
      counters: Object.fromEntries(this.counters.entries()),
      events: this.events.slice(),
    };
  }

  exportDiagnostics() {
    this.record({ name: "support_export", outcome: "success", attributes: { source: "local" } });
    let snapshot = this.snapshot();
    let serialized = JSON.stringify(snapshot);
    while (serializedBytes(serialized) > ROOTINE_DIAGNOSTIC_EXPORT_MAX_BYTES && snapshot.events.length > 0) {
      snapshot = { ...snapshot, events: snapshot.events.slice(1) };
      serialized = JSON.stringify(snapshot);
    }
    return serialized;
  }

  reset() {
    this.events = [];
    this.counters.clear();
  }
}

export const rootineObservability = RootineObservability.shared;

File: ios/Rootine/Rootine/Core/Observability.swift
import Foundation

enum RootineTelemetryOutcome: String, Codable, Sendable {
    case success
    case failure
    case degraded
    case cancelled
    case unknown
}

struct RootineDiagnosticEvent: Codable, Equatable, Sendable {
    let schemaVersion: Int
    let name: String
    let outcome: RootineTelemetryOutcome
    let at: String
    let durationMilliseconds: Int?
    let correlationID: String?
    let operationID: String?
    let attributes: [String: String]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case name
        case outcome
        case at
        case durationMilliseconds = "duration_ms"
        case correlationID = "correlation_id"
        case operationID = "operation_id"
        case attributes
    }
}

struct RootineDiagnosticSnapshot: Codable, Equatable, Sendable {
    let schemaVersion: Int
    let generatedAt: String
    let supportID: String
    let counters: [String: Int]
    let events: [RootineDiagnosticEvent]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case generatedAt = "generated_at"
        case supportID = "support_id"
        case counters
        case events
    }
}

enum RootineHealthCounter: String, CaseIterable, Sendable {
    case authSuccess = "auth_success"
    case authFailure = "auth_failure"
    case syncPullSuccess = "sync_pull_success"
    case syncPullFailure = "sync_pull_failure"
    case syncPushSuccess = "sync_push_success"
    case syncPushFailure = "sync_push_failure"
    case syncRetry = "sync_retry"
    case syncConflict = "sync_conflict"
    case syncCursorExpired = "sync_cursor_expired"
    case syncUnauthorized = "sync_unauthorized"
    case realtimeConnected = "realtime_connected"
    case realtimeReconnect = "realtime_reconnect"
    case realtimeFailure = "realtime_failure"
    case qrDetected = "qr_detected"
    case qrSuccess = "qr_success"
    case qrFailure = "qr_failure"
    case apnsDelivered = "apns_delivered"
    case apnsFailed = "apns_failed"
    case apnsUnregistered = "apns_unregistered"
    case apnsRetry = "apns_retry"
    case materializerQuarantine = "materializer_quarantine"
    case deviceRegistered = "device_registered"
    case deviceRegistrationFailure = "device_registration_failure"
    case crashCaptured = "crash_captured"
}

/// Privacy-safe, vendor-neutral diagnostics. This is an in-process boundary;
/// callers can later provide an adapter without changing feature contracts.
final class RootineObservability: @unchecked Sendable {
    static let shared = RootineObservability()

    static let schemaVersion = 1
    static let maximumEventBytes = 4_096
    static let maximumExportBytes = 64 * 1_024
    static let maximumEvents = 64

    private static let safeAttributeKeys: Set<String> = [
        "action", "endpoint", "environment", "error", "entity", "entity_id", "format",
        "http_status", "permission", "provider", "reason", "source", "status", "trigger",
        "attempt", "batch_size", "change_count", "cursor", "revision", "retry_after_seconds",
        "queue_depth"
    ]

    private let lock = NSLock()
    private let maxEvents: Int
    private let now: () -> Date
    private(set) var supportID: String
    private var counters: [String: Int] = [:]
    private var events: [RootineDiagnosticEvent] = []

    init(
        supportID: String? = nil,
        maxEvents: Int = RootineObservability.maximumEvents,
        now: @escaping () -> Date = Date.init
    ) {
        self.supportID = supportID ?? "ios-\(UUID().uuidString.lowercased())"
        self.maxEvents = max(1, min(RootineObservability.maximumEvents, maxEvents))
        self.now = now
    }

    func increment(_ counter: RootineHealthCounter, amount: Int = 1) {
        guard amount > 0 else { return }
        lock.lock()
        counters[counter.rawValue] = min(Int.max, (counters[counter.rawValue] ?? 0) + min(amount, 1_000))
        lock.unlock()
    }

    @discardableResult
    func record(
        name: String,
        outcome: RootineTelemetryOutcome = .unknown,
        duration: TimeInterval? = nil,
        correlationID: String? = nil,
        operationID: String? = nil,
        attributes: [String: String] = [:]
    ) -> RootineDiagnosticEvent {
        let event = RootineDiagnosticEvent(
            schemaVersion: Self.schemaVersion,
            name: bounded(name, maximum: 64),
            outcome: outcome,
            at: bounded(ISO8601DateFormatter().string(from: now()), maximum: 48),
            durationMilliseconds: duration.map { max(0, Int(($0 * 1_000).rounded())) },
            correlationID: correlationID.map { bounded($0, maximum: 180) },
            operationID: operationID.map { bounded($0, maximum: 180) },
            attributes: safeAttributes(attributes)
        )
        let boundedEvent = boundedEvent(event)
        lock.lock()
        events.append(boundedEvent)
        if events.count > maxEvents { events.removeFirst(events.count - maxEvents) }
        lock.unlock()
        return boundedEvent
    }

    func recordAuth(outcome: RootineTelemetryOutcome, provider: String? = nil, error: String? = nil) {
        increment(outcome == .success ? .authSuccess : .authFailure)
        var attributes = ["provider": provider ?? "unknown"]
        if let error { attributes["error"] = errorCode(error) }
        record(name: "auth_outcome", outcome: outcome, attributes: attributes)
    }

    func recordSync(
        endpoint: String,
        outcome: RootineTelemetryOutcome,
        duration: TimeInterval? = nil,
        status: Int? = nil,
        correlationID: String? = nil,
        operationID: String? = nil,
        trigger: String? = nil,
        error: String? = nil,
        attributes: [String: String] = [:]
    ) {
        let isSuccess = outcome == .success
        if endpoint == "pull" || endpoint == "bootstrap" { increment(isSuccess ? .syncPullSuccess : .syncPullFailure) }
        if endpoint == "push" { increment(isSuccess ? .syncPushSuccess : .syncPushFailure) }
        if outcome == .degraded { increment(.syncRetry) }
        if status == 401 || status == 403 { increment(.syncUnauthorized) }
        if status == 409 { increment(.syncCursorExpired) }
        var eventAttributes = attributes
        eventAttributes["endpoint"] = endpoint
        eventAttributes["status"] = status.map(String.init) ?? eventAttributes["status"] ?? "unknown"
        if let trigger { eventAttributes["trigger"] = trigger }
        if let error { eventAttributes["error"] = errorCode(error) }
        record(name: "sync_operation", outcome: outcome, duration: duration, correlationID: correlationID, operationID: operationID, attributes: eventAttributes)
    }

    func recordQR(outcome: RootineTelemetryOutcome, format: String = "qr", error: String? = nil) {
        increment(.qrDetected)
        switch outcome {
        case .success: increment(.qrSuccess)
        case .failure: increment(.qrFailure)
        default: break
        }
        var attributes = ["format": safeScanFormat(format)]
        if let error { attributes["error"] = errorCode(error) }
        record(name: "qr_scan", outcome: outcome, attributes: attributes)
    }

    func recordQRDetected(format: String) {
        increment(.qrDetected)
        record(name: "qr_scan", outcome: .unknown, attributes: ["format": safeScanFormat(format)])
    }

    func recordNotificationDelivery(status: String, retryable: Bool = false, responseCode: Int? = nil) {
        switch status {
        case "delivered": increment(.apnsDelivered)
        case "unregistered": increment(.apnsUnregistered)
        default: increment(.apnsFailed)
        }
        if retryable { increment(.apnsRetry) }
        var attributes = ["status": status]
        if let responseCode { attributes["http_status"] = String(responseCode) }
        record(name: "notification_delivery", outcome: status == "delivered" ? .success : retryable ? .degraded : .failure, attributes: attributes)
    }

    func recordMaterializerQuarantine(reason: String? = nil) {
        increment(.materializerQuarantine)
        var attributes: [String: String] = [:]
        if let reason { attributes["reason"] = errorCode(reason) }
        record(name: "materializer_quarantine", outcome: .degraded, attributes: attributes)
    }

    func recordDeviceHealth(outcome: RootineTelemetryOutcome, permission: String? = nil, environment: String? = nil, error: String? = nil) {
        increment(outcome == .success ? .deviceRegistered : .deviceRegistrationFailure)
        var attributes: [String: String] = [:]
        if let permission { attributes["permission"] = permission }
        if let environment { attributes["environment"] = environment }
        if let error { attributes["error"] = errorCode(error) }
        record(name: "device_health", outcome: outcome, attributes: attributes)
    }

    func recordCrash(error: String? = nil) {
        increment(.crashCaptured)
        var attributes: [String: String] = [:]
        if let error { attributes["error"] = errorCode(error) }
        record(name: "crash", outcome: .failure, attributes: attributes)
    }

    func installCrashReporter() {
        NSSetUncaughtExceptionHandler { _ in
            RootineObservability.shared.recordCrash(error: "uncaught_exception")
        }
    }

    func snapshot() -> RootineDiagnosticSnapshot {
        lock.lock()
        let snapshot = RootineDiagnosticSnapshot(
            schemaVersion: Self.schemaVersion,
            generatedAt: ISO8601DateFormatter().string(from: now()),
            supportID: supportID,
            counters: counters,
            events: events
        )
        lock.unlock()
        return snapshot
    }

    func exportDiagnostics() -> Data {
        record(name: "support_export", outcome: .success, attributes: ["format": "json"])
        var snapshotValue = snapshot()
        var serialized = encode(snapshotValue)
        while serialized.count > Self.maximumExportBytes && !snapshotValue.events.isEmpty {
            snapshotValue = RootineDiagnosticSnapshot(
                schemaVersion: snapshotValue.schemaVersion,
                generatedAt: snapshotValue.generatedAt,
                supportID: snapshotValue.supportID,
                counters: snapshotValue.counters,
                events: Array(snapshotValue.events.dropFirst())
            )
            serialized = encode(snapshotValue)
        }
        return serialized
    }

    func reset() {
        lock.lock()
        counters.removeAll()
        events.removeAll()
        lock.unlock()
    }

    private func safeAttributes(_ input: [String: String]) -> [String: String] {
        input
            .filter { Self.safeAttributeKeys.contains($0.key) }
            .prefix(24)
            .reduce(into: [String: String]()) { result, item in
                result[item.key] = item.key == "error" || item.key == "reason" ? errorCode(item.value) : bounded(item.value)
            }
    }

    private func boundedEvent(_ event: RootineDiagnosticEvent) -> RootineDiagnosticEvent {
        guard encode(event).count > Self.maximumEventBytes else { return event }
        return RootineDiagnosticEvent(
            schemaVersion: event.schemaVersion,
            name: event.name,
            outcome: event.outcome,
            at: event.at,
            durationMilliseconds: event.durationMilliseconds,
            correlationID: event.correlationID,
            operationID: event.operationID,
            attributes: [:]
        )
    }

    private func errorCode(_ value: String?) -> String {
        let normalized = (value ?? "").lowercased()
        if normalized.contains("unauthor") || normalized.contains("401") { return "unauthorized" }
        if normalized.contains("cursor") && normalized.contains("expir") { return "cursor_expired" }
        if normalized.contains("conflict") || normalized.contains("revision") { return "conflict" }
        if normalized.contains("rate") || normalized.contains("429") { return "rate_limited" }
        if normalized.contains("timeout") || normalized.contains("abort") { return "timeout" }
        if normalized.contains("network") || normalized.contains("fetch") { return "network" }
        if normalized.contains("invalid") || normalized.contains("schema") { return "invalid" }
        if normalized.contains("server") || normalized.contains("500") { return "server" }
        return "unknown"
    }

    private func safeScanFormat(_ value: String) -> String {
        value == "qr" || value == "barcode" ? value : "unknown"
    }

    private func bounded(_ value: String, maximum: Int = 180) -> String {
        let normalized = value.unicodeScalars.map { $0.value < 32 || $0.value == 127 ? " " : String($0) }.joined().trimmingCharacters(in: .whitespacesAndNewlines)
        guard normalized.count > maximum else { return normalized }
        return String(normalized.prefix(maximum - 3)) + "..."
    }

    private func encode<T: Encodable>(_ value: T) -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return (try? encoder.encode(value)) ?? Data("{}".utf8)
    }
}
